### PR TITLE
Implement Interval Bound Propagation Training

### LINKED ISSUE #2037: Implementation of certified training via IBP
**Is your feature request related to a problem? Please describe.**
Currently, ART can support a few certified training schemes (de-randomized smoothing, zonotopes, and randomised smoothing). However IBP can provide excellent certified performance at a much lower computational overhead than, for example, zonotopes.

**Describe the solution you'd like**
Implementation of an IBP trainer in a similar style to the zonotope trainer. 

Paper Link: https://arxiv.org/abs/1810.12715

### PR DESCRIPTION
**Pull request by GiulioZizzo**
# Description

In this PR we include adversarial training with interval bound propagation. With careful training this can achieve excellent certified robustness.

Fixes #2037

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Tests can be found in test_interval.py. We check that the certification gives the correct results, runs without errors, propagates gradients correctly and when running without bounds gives the same results as regular neural networks.

**Test Configuration**:
- OS: MacOS
- Python version: 3.8
- 1.13.1
- Pytorch 1.13

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/defences/trainer/__init__.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from art.defences.trainer.trainer import Trainer
 from art.defences.trainer.adversarial_trainer import AdversarialTrainer
 from art.defences.trainer.certified_adversarial_trainer_pytorch import AdversarialTrainerCertifiedPytorch
+from art.defences.trainer.ibp_certified_trainer_pytorch import AdversarialTrainerCertifiedIBPPytorch
```
Comment (beat-buesser) [inline art/defences/trainer/__init__.py:7]:
```suggestion
from art.defences.trainer.ibp_certified_trainer_pytorch import AdversarialTrainerCertifiedIBPPyTorch
```

File: art/estimators/certification/interval/pytorch.py [hunk 2]
```diff
@@ -195,6 +206,13 @@ class PyTorchIBPClassifier(PyTorchIntervalBounds, PyTorchClassifier):
     to then verify if it can have its class changed given a certain perturbation.
 
     | Paper link: https://ieeexplore.ieee.org/document/8418593
+
+    This classifier has 3 modes which can be set via: classifier.model.set_forward_mode('mode')
+
+    'mode' can be one of:
+        + 'abstract': When we wish to certifiy datapoints and have abstract predictions
+        + 'concrete': When normal predictions need to be made
+        + 'attack': When we are interfacing with an ART attack (for example PGD).
```
Comment (beat-buesser) [inline art/estimators/certification/interval/pytorch.py:215]: Does this docstring build correctly?

File: art/estimators/certification/interval/pytorch.py [hunk 3]
```diff
@@ -184,9 +187,17 @@ def set_forward_mode(self, mode: str) -> None:
 
         :param mode: either concrete or abstract signifying how to run the forward pass
         """
-        assert mode in {"concrete", "abstract"}
+        assert mode in {"concrete", "abstract", "attack"}
         self.forward_mode = mode
 
+    def re_convert(self) -> None:
+        """
+        After an update on the convolutional weights, we re-convert them into the equivalent dense layer
```
Comment (beat-buesser) [inline art/estimators/certification/interval/pytorch.py:195]:
```suggestion
        After an update on the convolutional weights, re-convert weights into the equivalent dense layer.
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+    | Paper link: https://arxiv.org/pdf/1810.12715.pdf
+"""
+import logging
+import random
+import sys
+
+import numpy as np
+
+from sklearn.utils import shuffle
+from tqdm import tqdm
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:31]:
```suggestion
import numpy as np
from sklearn.utils import shuffle
from tqdm import tqdm
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
+# MIT License
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:1]: Please update `docs` and check if new docstrings build correctly.

File: art/estimators/certification/interval/pytorch.py [hunk 6]
```diff
@@ -376,3 +394,37 @@ def get_accuracy(preds: Union[np.ndarray, "torch.Tensor"], labels: Union[np.ndar
...
+
+    def concrete_loss(self, output: "torch.Tensor", target: "torch.Tensor") -> "torch.Tensor":
+        """
+        Access function to get the classifier loss
+
+        :param output: model predictions
+        :param target: ground truth labels
+
+        :return: loss value
+        """
```
Comment (beat-buesser) [inline art/estimators/certification/interval/pytorch.py:406]:
```suggestion
        :param output: model predictions
        :param target: ground truth labels
        :return: loss value
        """
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+        Grow the value by self.step_per_epoch
+        """
+        self.step_count += 1
+        if self.step_count > self.warmup and self.val < self.final_val:
+            self.val += self.step_per_epoch
+            self.val = min(self.val, self.final_val)
+        return self.val
+
+
+class AdversarialTrainerCertifiedIBPPytorch(Trainer):
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:98]:
```suggestion
class AdversarialTrainerCertifiedIBPPyTorch(Trainer):
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+
+    @staticmethod
+    def initialise_default_scheduler(initial_val: float, final_val: float, epochs: int) -> DefaultLinearScheduler:
+        """
+        Create linear schedulers based on default example values.
+
+        :param initial_val: initial value to begin the scheduler from
+        :param final_val: final value to end the scheduler at
+        :param epochs: total number of epochs
+        """
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:198]: Please add return value.

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+                   where axis=1 corresponds to the [lower, upper] bounds.
+
+                2. Or in regular concrete form, in which case the bounds/limits need to be supplied.
+        :param is_interval: if the datapoint is already in the correct interval format.
+        :param bounds: The perturbation range.
+        :param limits: The clipping to apply to the interval data.
+        :param batch_size: batch size to use when looping through the data
+        """
+
+        if self._classifier._model._model.forward_mode != "abstract":  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:449]: Is this required, it should be possible to do
```suggestion
        if self.classifier.model.forward_mode != "abstract":
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+            )
+        return self._classifier.predict_intervals(x, is_interval, bounds, limits, batch_size, **kwargs)
+
+    def set_forward_mode(self, mode: str) -> None:
+        """
+        Helper function to set the forward mode of the model
+
+        :param mode: either concrete or abstract signifying how to run the forward pass
+        """
+        self._classifier._model._model.set_forward_mode(mode)  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:462]:
```suggestion
        self.classifier.model.set_forward_mode(mode)
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+
+    def predict(self, x: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Perform prediction using the adversarially trained classifier.
+
+        :param x: Input samples.
+        :param kwargs: Other parameters to be passed on to the `predict` function of the classifier.
+        :return: Predictions for test set.
+        """
+        if self._classifier._model._model.forward_mode != "concrete":  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:417]:
```suggestion
        if self.classifier.model.forward_mode != "concrete":
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+        from art.estimators.certification.interval.pytorch import PyTorchIBPClassifier
+
+        if not isinstance(classifier, PyTorchIBPClassifier):
+            raise ValueError(
+                "The classifier to pass in should be of type PyTorchIBPClassifier which can be found in "
+                "art.estimators.certification.interval.pytorch.PyTorchIBPClassifier"
+            )
+
+        super().__init__(classifier=classifier)
+        self._classifier: "IBP_CERTIFIER_TYPE"
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:155]: Is the `_` required here?

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+                if self._classifier._use_amp:  # pragma: no cover # pylint: disable=W0212
+                    from apex import amp  # pylint: disable=E0611
+
+                    with amp.scale_loss(loss, self._classifier._optimizer) as scaled_loss:  # pylint: disable=W0212
+                        scaled_loss.backward()
+
+                else:
+                    loss.backward()
+
+                self._classifier._optimizer.step()  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:394]:
```suggestion
                self.classifier.optimizer.step()
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+                self._classifier.re_convert()
+
+                if verbose:
+                    pbar.set_description(
+                        f"Bound {bound:.2f}: "  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:399]: Is this in the correct line?
```suggestion
                        f"Bound {bound:.2f}: "
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+                )  # pylint: disable=W0212
+                epoch_non_cert_loss.append(non_cert_loss)
+                non_cert_acc.append(acc)
+
+                loss = certified_loss * loss_weighting_k + non_cert_loss * (1 - loss_weighting_k)
+                # Do training
+                if self._classifier._use_amp:  # pragma: no cover # pylint: disable=W0212
+                    from apex import amp  # pylint: disable=E0611
+
+                    with amp.scale_loss(loss, self._classifier._optimizer) as scaled_loss:  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:388]:
```suggestion
                    with amp.scale_loss(loss, self.classifier.optimizer) as scaled_loss:
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+
+            cert_loss = []
+            cert_acc = []
+
+            for m in pbar:
+                x_batch = np.copy(x_cert[batch_size * m : batch_size * (m + 1)])
+                y_batch = np.copy(y_cert[batch_size * m : batch_size * (m + 1)])
+
+                # Zero the parameter gradients
+                self._classifier._optimizer.zero_grad()  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:326]:
```suggestion
                self.classifier.optimizer.zero_grad()
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+        if limits is None:
+            raise ValueError(
+                "Please provide values for the clipping limits of the data. "
+                "Typical images will have limits of [0.0, 1.0]. "
+            )
+
+        # Set model mode
+        self._classifier._model.train(mode=training_mode)  # pylint: disable=W0212
+
+        if self._classifier._optimizer is None:  # pragma: no cover # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:268]:
```suggestion
        if self.classifier.optimizer is None:
```

File: art/defences/trainer/ibp_certified_trainer_pytorch.py [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+            raise ValueError("Value of `epochs` not defined.")
+
+        if limits is None:
+            raise ValueError(
+                "Please provide values for the clipping limits of the data. "
+                "Typical images will have limits of [0.0, 1.0]. "
+            )
+
+        # Set model mode
+        self._classifier._model.train(mode=training_mode)  # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/ibp_certified_trainer_pytorch.py:266]:
```suggestion
        self.classifier.model.train(mode=training_mode)
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @GiulioZizzo Thank you very much for your pull request in Interval Bound Propagation training! I have left a few comments, what do you think?

File: art/defences/trainer/ibp_certified_trainer_pytorch.py (hunk 14 above)
Comment (GiulioZizzo) [reply art/defences/trainer/ibp_certified_trainer_pytorch.py]: Hi Beat, we removed the need for this warning, but yes oddly enough this was the line despite the warning actually occurring further down. It seemed to be that it needed to be placed when the string was first opened .

File: art/defences/trainer/ibp_certified_trainer_pytorch.py (hunk 9 above)
Comment (GiulioZizzo) [reply art/defences/trainer/ibp_certified_trainer_pytorch.py]: I agree! Change implemented.

File: art/defences/trainer/ibp_certified_trainer_pytorch.py (hunk 11 above)
Comment (GiulioZizzo) [reply art/defences/trainer/ibp_certified_trainer_pytorch.py]: Agree! Change Implemented

File: art/defences/trainer/ibp_certified_trainer_pytorch.py (hunk 12 above)
Comment (GiulioZizzo) [reply art/defences/trainer/ibp_certified_trainer_pytorch.py]: Changed with additional method in the Trainer class.

File: art/defences/trainer/ibp_certified_trainer_pytorch.py (hunk 13 above)
Comment (GiulioZizzo) [reply art/defences/trainer/ibp_certified_trainer_pytorch.py]: Changed with additional method in Trainer class

File: art/defences/trainer/__init__.py (hunk 1 above)
Comment (GiulioZizzo) [reply art/defences/trainer/__init__.py:7]: Agree, name changed!

File: art/estimators/certification/interval/pytorch.py (hunk 2 above)
Comment (GiulioZizzo) [reply art/estimators/certification/interval/pytorch.py:215]: checked, all ok!

**Review by beat-buesser — APPROVED** [review]
